### PR TITLE
scm-webhook

### PR DESCRIPTION
**Pull request by supritat**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### CODE REVIEW

File: jazz_scm-webhook/index.js [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+		
+		eventBody = event.body;
+
+		if(scmSource === 'bitbucket'){
+			userName = (eventBody.actor.username) ? eventBody.actor.username : '';
+			eventKey = event.headers['X-Event-Key'];
+			service = eventBody.repository.slug;
+			repositoryLink = eventBody.repository.links.self[0].href;
+			bitbucketScmContext(eventKey, eventBody, config)
+			.then((res) => updateEvents(res, service, userName,repositoryLink, config))
```
Comment (suryajak) [inline jazz_scm-webhook/index.js]: We should push the bitbucket code (extracting values from event and eventBody to bitbucket specific function. We should do similar thing for gitlabs.  
The resultant code can be like 

```
return new Promise((resolve, reject) => { 
      getScmContextDetails(scmSource, eventKey, eventBody, config)
      .then ((scmDetails) => updateEventsWithScmDetails(scmDetails)
      .then(function(res) { resolve(res); }
      .catch ...
}
```

where getScmContextDetails has the fork between `bitbucket` and `gitlab`
``` 
function getScmContextDetails (scmSource, eventKey, eventBody, config){
if(scmSource === 'bitbucket'){
} else if(scmSource === 'gitlab'){

}
```

File: jazz_scm-webhook/config/dev-config.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
+{
+  "EVENTS_URL": "/dev/jazz/events",
+  "SERVICE_API_URL" : "https://{conf-apikey}.execute-api.{conf-region}.amazonaws.com",
+  "TOKEN_URL" : "/dev/jazz/login",
+  "EVENT_STATUS": {
+    "completed": "COMPLETED"
+  },
+  "EVENT_HANDLER": {
+    "bitbucket": "BITBUCKET",
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/config/dev-config.json]: Updated with SCM_TYPE

File: jazz_scm-webhook/index.js (hunk 1 above)
Comment (supritat) [reply jazz_scm-webhook/index.js]: Updated the changes

File: jazz_scm-webhook/config/dev-config.json [hunk 3]
```diff
@@ -1,64 +1,39 @@
...
     "COMMIT_CODE",
     "DELETE_BRANCH",
+    "DELETE_TAG",
     "RAISE_PR",
     "MERGE_PR",
     "DECLINE_PR",
     "UPDATE_PR",
-    "COMMENT_PR"
+    "COMMENT_PR",
+    "DELETE_TAG"
```
Comment (suryajak) [inline jazz_scm-webhook/config/dev-config.json]: Duplicate event name

File: jazz_scm-webhook/config/stg-config.json [hunk 4]
```diff
@@ -1,64 +1,39 @@
...
     "COMMIT_CODE",
     "DELETE_BRANCH",
+    "DELETE_TAG",
     "RAISE_PR",
     "MERGE_PR",
     "DECLINE_PR",
     "UPDATE_PR",
-    "COMMENT_PR"
+    "COMMENT_PR",
+    "DELETE_TAG"
```
Comment (suryajak) [inline jazz_scm-webhook/config/stg-config.json]: duplicate event name

File: jazz_scm-webhook/index.js [hunk 5]
```diff
@@ -147,44 +146,233 @@ module.exports.handler = (event, context, cb) => {
...
+						resolve(result);
 					} 		
-				} else if(changes && !changes.created && !changes.closed && changes.old){
+				} else if(changes && !changes.created && !changes.closed && changes.old ){
 					result.branch = changes.new.name;
 					result.hash = changes.new.target.hash;
 					result.event_name = 'COMMIT_CODE';
+					resolve(result);
+				} else {
+					reject("Invalid push event")
```
Comment (suryajak) [inline jazz_scm-webhook/index.js]: missing semi-colon

File: jazz_scm-webhook/index.js [hunk 6]
```diff
@@ -147,44 +146,233 @@ module.exports.handler = (event, context, cb) => {
...
 			case 'pullrequest:updated':
 				result.event_name = 'UPDATE_PR';
+				resolve(result);
 				break;
 			case 'pullrequest:comment':
 				result.event_name = 'COMMENT_PR';
+				resolve(result);
 				break;
+			default:
+				reject("Invalid event key")
```
Comment (suryajak) [inline jazz_scm-webhook/index.js]: missing semi-colon

File: jazz_scm-webhook/config/dev-config.json (hunk 3 above)
Comment (supritat) [reply jazz_scm-webhook/config/dev-config.json]: Removed duplication

File: jazz_scm-webhook/config/stg-config.json (hunk 4 above)
Comment (supritat) [reply jazz_scm-webhook/config/stg-config.json]: Removed duplication

File: jazz_scm-webhook/index.js (hunk 5 above)
Comment (supritat) [reply jazz_scm-webhook/index.js]: Fixed

File: jazz_scm-webhook/index.js (hunk 6 above)
Comment (supritat) [reply jazz_scm-webhook/index.js]: Fixed

File: jazz_scm-webhook/index.js [hunk 7]
```diff
@@ -147,44 +146,233 @@ module.exports.handler = (event, context, cb) => {
...
+		}		
+	});
+}
+
+function gitlabScmContextDetailsDetails(eventKey, body, config){
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/index.js]: Fixed

File: jazz_scm-webhook/index.js [hunk 8]
```diff
@@ -147,44 +146,233 @@ module.exports.handler = (event, context, cb) => {
...
-		return result;
-	}
-};
+	})
+}
+
+function updateEventsWithScmDetails(servObj, config){
+	logger.info("Inside updateEventsWithScmDetails: "+ JSON.stringify(servObj));
+	return new Promise((resolve, reject) => {
+		var serviceName = (servObj.service) ? servObj.service.split("_")[1] : "",
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/index.js]: Done as mentioned

File: jazz_scm-webhook/index.js [hunk 9]
```diff
@@ -147,44 +146,233 @@ module.exports.handler = (event, context, cb) => {
...
+						'Authorization': authToken
+					},
+					body : bodyObj,
+					json: true,
+					rejectUnauthorized: false
+				};
+				request(options, function (error, response, body) {
+					if (error) {
+						logger.error('Error invoking service: ' + JSON.stringify(error));
+						reject(error)
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/index.js]: Fixed

File: jazz_scm-webhook/index.js [hunk 10]
```diff
@@ -147,44 +146,233 @@ module.exports.handler = (event, context, cb) => {
...
+					} else {
+						if(response.statusCode === 200){
+							var output = {
+								message: 'successfully recorded git activity to jazz_events.',
+								event_id : body.data.event_id
+							};
+							resolve(responseObj(output, body.input));
+						} else {
+							logger.error("StatusCode :"+ JSON.stringify(response));
+							reject(response.body.message)
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/index.js]: Fixed

File: jazz_scm-webhook/index.js [hunk 11]
```diff
@@ -147,44 +146,233 @@ module.exports.handler = (event, context, cb) => {
...
+						}
+					}
+				});
+			})
+			.catch(function(err){
+				logger.error(err)
+				reject(err)
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/index.js]: Fixed

File: jazz_scm-webhook/package.json [hunk 12]
```diff
@@ -1,10 +1,12 @@
 {
-  "name": "webhook-api",
+  "name": "test_webhook-api",
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/package.json]: Updated with jazz_scm-webhook

File: jazz_scm-webhook/swagger/swagger.json [hunk 13]
```diff
@@ -1,559 +1,560 @@
...
-                  "description": "List all repository links",
-                  "items": {
+            }
+        },
+        "InputObj": {
+            "type": "object",
+            "properties": {
+                "actor": {
+                    "type": "object",
+                    "description": "Bitbucket user details",
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/swagger/swagger.json]: Updated

File: jazz_scm-webhook/swagger/swagger.json [hunk 14]
```diff
@@ -1,559 +1,560 @@
...
+                    "properties": {
+                        "username": {
+                            "type": "string",
+                            "description": "User name of the bitbucket user"
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/swagger/swagger.json]: Updated

File: jazz_scm-webhook/swagger/swagger.json [hunk 15]
```diff
@@ -1,559 +1,560 @@
...
+                        },
+                        "displayName": {
+                            "type": "string",
+                            "description": "Display name of the bitbucket user"
```
[the remark this answers is not in the document]
Comment (supritat) [reply jazz_scm-webhook/swagger/swagger.json]: Updated

File: jazz_scm-webhook/components/config.js [hunk 16]
```diff
@@ -22,20 +23,19 @@
...
-  }else {
-      stage = event.stage
-  } 
-  
-  var configObj;
-  
-  if (stage) {
-      configObj = require(`../config/${stage}-config.json`);
-  } 
+  var stage = event.stage;
```
Comment (svidhani) [inline jazz_scm-webhook/components/config.js]: Suggestion: lines 26-38 can be reduced by using a hashmap. the performance here is better than comparing if else and the code can be concise. 

stage = { 'dev': 'dev', 'stg': 'stg', 'prod': 'prod'}
var stageConfig = stage[event.stage] ? stage[event.stage] : event;
var configObj = require(`../config/${stageConfig}-config.json`);

File: jazz_scm-webhook/components/config.js [hunk 17]
```diff
@@ -22,20 +23,10 @@
...
-  }else {
-      stage = event.stage
-  } 
-  
-  var configObj;
-  
-  if (stage) {
-      configObj = require(`../config/${stage}-config.json`);
-  } 
+  var stage = event.stage,
```
Comment (suryajak) [inline jazz_scm-webhook/components/config.js]: I had condensed this code to [this](https://github.com/tmobile/jazz/blob/master/jazz_usermanagement/components/config.js#L23). If we are changing this for performance improvements, the one [here](https://github.com/tmobile/jazz/blob/master/jazz_usermanagement/components/config.js) is perhaps more optimized. @svidhani: What do you think?

Comment (svidhani) [reply jazz_scm-webhook/components/config.js]: even better :). thanks

**Pull request merged**